Makefile

```diff
@@ -1,231 +1,293 @@
+LDFLAGS_EXTRAS=
+LDFLAGS=$(LDFLAGS_PROJ_VERSION) $(LDFLAGS_UPSTREAM_VERSION) $(LDFLAGS_EXTRAS) $(LDFLAGS_STRIP_SYMBOLS)
+
+# Create a `.make` directory for tracking targets which don't generate a single file output. This should be ignored by git.
+# For targets which either don't generate a single file output, or the output file is committed, we use a "sentinel"
+# file within `.make/` to track the staleness of the target and only rebuild when needed.
+# For each phony target, we create an internal target with the same name, but prefixed with `.make/` where the work is performed.
+# At the end of each internal target we run `@touch $@` to update the file which is the name of the target.
+
+# Ensure all directories exist before evaluating targets to avoid issues with `touch` creating directories.
+_ := $(shell mkdir -p .make bin .pulumi/bin)
+
+# Installs all necessary tools with mise and records completion in a sentinel
+# file so dependent targets can participate in make's caching behaviour. The
+# environment is refreshed via an order-only prerequisite so it still runs on
+# every invocation without invalidating the sentinel.
+mise_install: .make/mise_install | mise_env
+
+.PHONY: mise_env
+mise_env:
+	@mise env -q  > /dev/null
+
+.make/mise_install:
+	@mise install -q
+	@touch $@
+
+# Build the provider and all SDKs and install ready for testing
+build: .make/mise_install provider build_sdks install_sdks
+build: | mise_env
+# Keep aliases for old targets to ensure backwards compatibility
+development: build
 only_build: build
+# Prepare the workspace for building the provider and SDKs
+# Importantly this is run by CI ahead of restoring the bin directory and resuming SDK builds
+prepare_local_workspace: .make/mise_install upstream
+prepare_local_workspace: | mise_env
+# Creates all generated files which need to be committed
+generate: generate_sdks schema
+generate_sdks: generate_nodejs generate_python generate_dotnet generate_go generate_java
+build_sdks: build_nodejs build_python build_dotnet build_go build_java
+install_sdks: install_nodejs_sdk install_python_sdk install_dotnet_sdk install_go_sdk install_java_sdk
+.PHONY: development only_build build generate generate_sdks build_sdks install_sdks mise_install mise_env
 
-only_tfgen: install_plugins upstream build_schema
-tfgen: only_tfgen generate_examples
-
-# Generate examples after the schema is generated
-generate_examples: examples build_schema_post_examples
-
-# Build the tfgen binary and generate the schema
-build_schema build_schema_post_examples: $(PULUMICTL_BIN)
-build_schema build_schema_post_examples:
-	(cd provider && go build -o $(WORKING_DIR)/bin/${TFGEN} -ldflags "-X ${PROJECT}/${VERSION_PATH}=${VERSION}" ${PROJECT}/${PROVIDER_PATH}/cmd/${TFGEN})
-	$(WORKING_DIR)/bin/${TFGEN} schema --out provider/cmd/${PROVIDER}
-	(cd provider && VERSION=$(VERSION) go generate cmd/${PROVIDER}/main.go)
-
-bin/pulumi-java-gen: .pulumi-java-gen.version $(PULUMICTL_BIN)
```

Comment:
Should bring back the old java gen logic for now and then upgrade to the bridge-native java generator in a follow-up PR.

Reply:
Given that we need to manually navigate the transition from pulumi-java-gen to pulumi-language-java regardless of what CI is in place, we might as well get #311 figured out first and then revisit this one after we've dropped pulumi-java-gen.